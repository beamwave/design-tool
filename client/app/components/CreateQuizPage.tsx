import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import fontawesome from '@fortawesome/fontawesome'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { startCreateNewQuiz } from '../actions/create'

interface IProps {
  history?: any
  username?: any
  quizzes?: any
  startCreateNewQuiz: (quiz: any, any) => any
}

export class CreateQuizPage extends Component<IProps, any> {
  state = {
    quiz: {
      author: this.props.username,
      title: '',
      description: '',
      isPrivate: true,
      type: 'quiz'
    },
    quizzes: this.props.quizzes
  }

  private updateQuiz = (e: any, arg1: string) => {
    switch (arg1) {
      case 'title':
        this.setState({
          quiz: {
            ...this.state.quiz,
            title: e.target.value
          }
        })
        break
      case 'description':
        this.setState({
          quiz: {
            ...this.state.quiz,
            description: e.target.value
          }
        })
      default:
        break
    }
    // console.log(this.state)
  }

  private updatePrivate = (e: any) => {
    switch (e.target.value) {
      case 'private':
        this.setState({
          quiz: {
            ...this.state.quiz,
            isPrivate: true
          }
        })
        break

      default:
        this.setState({
          quiz: {
            ...this.state.quiz,
            isPrivate: false
          }
        })
        break
    }
  }

  private submitQuiz = async (e: any) => {
    e.preventDefault()
    const { startCreateNewQuiz, username } = this.props
    const { quiz } = this.state

    // save quiz to db and update quizzes
    const createAction = await startCreateNewQuiz(quiz, this.props.history)

    // this.props.history.push('/quizzes/created')
    // this.props.history.push('/edit-quiz/' + uuid)
    // .then(res => {
    //   console.log('add quiz props')
    //   console.log(this.props)
    //   this.props.history.push('/add-question')
    // })
    // .catch(err =>
    //   console.log(
    //     'make sure the user knows that the quiz did not make it through'
    //   )
    // )
  }

  // @ts-ignore
  render = () => {
    const {
      quiz: { title }
    } = this.state
    return (
      <div className="create-quiz-page">
        <main>
          <div
            className={
              title.length > 0
                ? 'create-quiz-container valid'
                : 'create-quiz-container invalid'
            }
          >
            <p className="title">Create a quiz</p>
            <form className="details">
              <div className="group">
                <label htmlFor="create-quiz-title">Title</label>
                <input
                  type="text"
                  id="create-quiz-title"
                  value={this.state.quiz.title}
                  onChange={(e: any) => {
                    this.updateQuiz(e, 'title')
                  }}
                />
              </div>
              <div className="group">
                <label htmlFor="create-quiz-description">Description</label>
                <textarea
                  id="create-quiz-description"
                  value={this.state.quiz.description}
                  rows={3}
                  onChange={(e: any) => {
                    this.updateQuiz(e, 'description')
                  }}
                />
              </div>
              <div className="group">
                <label htmlFor="owner">Owner</label>
                <FontAwesomeIcon className="icon" icon="angle-down" />
                <select
                  name="owner"
                  id="question-options-dropdown"
                  onChange={this.updatePrivate}
                >
                  <option value="choose" hidden>
                    select
                  </option>
                  <option value="private">private</option>
                  <option value="public">public</option>
                </select>
              </div>
              <button
                type="button"
                className="save-quiz"
                onClick={this.submitQuiz}
                disabled={title.length > 0 ? false : true}
              >
                Create
              </button>
            </form>
          </div>
        </main>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  username: state.auth.username,
  quizzes: state.quizzes.all
})

export default connect<any, IProps>(
  mapStateToProps,
  { startCreateNewQuiz }
)(CreateQuizPage)
