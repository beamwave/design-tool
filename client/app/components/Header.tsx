import React, { Component } from 'react'
import { Link, withRouter } from 'react-router-dom'
import { connect } from 'react-redux'
import {
  Dropdown,
  DropdownToggle,
  DropdownMenu,
  DropdownItem
} from 'reactstrap'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

import SearchResults from './SearchResults'
import {
  sidebar,
  toggleEditMode,
  changeCategory,
  trainingWheelsProtocol,
  startSearch
} from '../actions/app'
import { startSearchByTag } from '../actions/images'
import { startLogout } from '../actions/auth'
import { loadModal } from '../actions/modal'
import { ADD_IMAGE_MODAL } from '../constants/modaltypes'

interface RProps {
  photo?: string
  username?: string
  isAuthenticated?: boolean
  trainingWheels?: boolean
  category?: string
  backButton?: boolean
  history?: any
}

interface RState {
  sidebar?: any
  startSearch?: (query: any) => any
  startSearchByTag?: (data: any) => any
  toggleEditMode?: () => any
  trainingWheelsProtocol?: () => any
  changeCategory?: (string) => any
  loadModal?: (string) => any
  startLogout?: () => void
}

type Props = RProps & RState

export class Header extends Component<Props> {
  state = {
    category: this.props.category,
    dropdownOpen: false,
    showSearch: false,
    term: '',
    editMode: false
  }

  toggleDropdown = () =>
    this.setState({ dropdownOpen: !this.state.dropdownOpen })

  toggleTrainingWheels = () => this.props.trainingWheelsProtocol()

  toggleEditMode = () => this.props.toggleEditMode()

  // needs to be attached to document for click detection
  componentDidMount = () =>
    document.addEventListener('click', this.determineClick, false)

  // needs to be attached to document for click detection
  componentWillUnmount = () =>
    document.removeEventListener('click', this.determineClick)

  addImageModal = () => this.props.loadModal(ADD_IMAGE_MODAL)

  handleCategoryChange = ({ target }) => {
    this.props.changeCategory(target.value)
  }

  determineClick = e => {
    // if click outside modal or not typing in input, hide search
    if (!e.target.closest('.section') && !e.target.dataset.search) {
      this.hideSearch()
    }

    // else if you click inside or type in input, ...
  }

  showSearch = () => this.setState({ showSearch: true })

  hideSearch = () => this.setState({ showSearch: false })

  handleSearch = async e => {
    // get search query
    const {
      target: { value: query }
    } = e

    const { category, startSearch } = this.props

    // this.setState({
    //   query
    // } as any)

    // send term to database
    await startSearch({ query, category })
  }

  handleSelectColor = category => {
    if (category === 'art') return 'select teal'
    if (category === 'comments') return 'select violet'
    if (category === 'clothes') return 'select blue'
    if (category === 'people') return 'select grey'
  }

  handleSubmit = e => {
    e.preventDefault()
    this.setState({
      term: this.state.term
    })
  }

  render = () => {
    const {
      photo,
      username,
      isAuthenticated,
      trainingWheels,
      category,
      startLogout,
      backButton,
      history
    } = this.props
    const { showSearch, term } = this.state

    return (
      <header className="nav-header">
        <SearchResults query={term} showSearch={showSearch} />
        {!backButton ? (
          <div className="plus" onClick={this.addImageModal}>
            <FontAwesomeIcon icon="plus" />
            <p className="hint">upload image</p>
          </div>
        ) : (
          <div className="plus" onClick={() => history.goBack()}>
            <FontAwesomeIcon icon="arrow-left" />
            <p className="hint">go back</p>
          </div>
        )}
        <div className="settings">
          <FontAwesomeIcon icon="cog" />
          <p className="hint">settings</p>
        </div>
        {!backButton && (
          <div className="protocol" onClick={this.toggleTrainingWheels}>
            {trainingWheels ? (
              <FontAwesomeIcon icon="lock" />
            ) : (
              <FontAwesomeIcon icon="unlock" />
            )}
            <p className="hint">
              {trainingWheels ? 'deactive' : 'active'} training wheels protocol
              )}
            </p>
          </div>
        )}
        {backButton && (
          <div className="protocol" onClick={this.toggleEditMode}>
            <FontAwesomeIcon icon="sliders-h" />
            <p className="hint">adjust image settings</p>
          </div>
        )}
        <div className="search-group">
          <FontAwesomeIcon icon="search" className="icon" />
          <form className="submit" onSubmit={this.handleSubmit}>
            <input
              type="text"
              className="input spawnSearch"
              placeholder="Search"
              // data-search={true}
              // onFocus={this.showSearch}
              onChange={this.handleSearch}
              value={term}
            />

            <div className="type-choice">
              <div className="angle-down">
                <FontAwesomeIcon icon="angle-down" className="icon" />
              </div>
              <select
                className={this.handleSelectColor(category)}
                value={category}
                onChange={this.handleCategoryChange}
              >
                <option value="art">art</option>
                <option value="comments">comments</option>
                <option value="clothes">clothes</option>
                <option value="people">people</option>
              </select>
            </div>
            <div className="search-options">
              <p className="save-search">save search</p>
              <p className="advanced-search">advanced search</p>
            </div>
          </form>
        </div>

        <Dropdown
          isOpen={this.state.dropdownOpen}
          toggle={this.toggleDropdown}
          className="dropdown-root"
        >
          <DropdownToggle className="dropdown-toggle">
            <p className="nav-username">{username}</p>
            <div
              className="image"
              style={{
                // url bg-pos: x% y% / bg-scale: %
                background: `url(${photo}) 50% 5% / 110% no-repeat`,
                marginRight: 15
              }}
            />
            <FontAwesomeIcon icon="angle-down" className="icon fa-angle-down" />
          </DropdownToggle>
          <DropdownMenu
            right
            className="dropdown-menu"
            style={{
              display: this.state.dropdownOpen === false ? 'none' : 'block'
            }}
          >
            <DropdownItem className="dropdown-item">
              <Link to="/profile" className="profile">
                Profile
              </Link>
            </DropdownItem>
            <DropdownItem
              className="dropdown-item"
              onClick={() => startLogout()}
            >
              {isAuthenticated ? <p className="logout">Logout</p> : null}
            </DropdownItem>
          </DropdownMenu>
        </Dropdown>
      </header>
    )
  }
}

const mapStateToProps = (state, props) => ({
  photo: state.auth.profileImage,
  username: state.auth.username,
  isAuthenticated: !!state.auth.token,
  category: state.app.category,
  trainingWheels: state.app.trainingWheels,
  backButton: props.match.params._id ? true : false // for some reason this works
})

export default withRouter<any>(
  connect<RProps, any>(
    mapStateToProps,
    {
      sidebar,
      toggleEditMode,
      changeCategory,
      trainingWheelsProtocol,
      startSearchByTag,
      startSearch,
      loadModal,
      startLogout
    }
  )(Header)
)
