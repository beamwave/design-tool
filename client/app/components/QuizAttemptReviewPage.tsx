import React, { Component } from 'react'
import moment from 'moment'
import ImageGallery from 'react-grid-gallery'
import numeral from 'numeral'
import { connect } from 'react-redux'
import { RouteProps } from 'react-router'
import { Link } from 'react-router-dom'
import {
  startAddAnswerToArray,
  changeQuestionNumber,
  addMultipleSelectAnswer,
  addMultipleChoiceAnswer,
  updateMultipleSelectAnswer,
  updateAnswerArray,
  submitQuizAttempt,
  clearQuizAttempt
} from '../actions/quizzes'
import { loadModal } from '../actions/modal'
import {
  SUBMIT_QUIZ_MODAL,
  REPORT_QUESTION_MODAL
} from '../constants/modaltypes'
import Spinner from 'react-spinkit'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

// interface IProps {
interface IProps extends RouteProps {
  username: any
  quiz: any
  quizAttempt: any
  questionNumber: any
  multipleSelectAnswer: any
  multipleChoiceAnswer: any[]
  answerArray: any
  history?: any
  quizResults: any
  clearQuizAttempt: (reset: number) => void
  loadModal: (any) => any
  changeQuestionNumber: (questionNumber: number) => void
  startAddAnswerToArray: (answerObj: {}) => void
  addMultipleChoiceAnswer: (answerObj: {}) => void
  addMultipleSelectAnswer: (answerObj: {}) => void
  updateMultipleSelectAnswer: (answerArray: any) => void
  handleSelection: (choice?: any, title?: any) => any
  updateAnswerArray: (answerArray: any[]) => void
  submitQuizAttempt: (
    quizUUID: string,
    user: string,
    attemptUUID: string,
    answerArray: any[]
  ) => void
}

const questionStyle = {
  cursor: 'pointer'
}

export class QuizAttemptReviewPage extends Component<IProps, any> {
  constructor(props) {
    super(props)
  }

  params = window.location.href.split('/')
  retakeIndex = this.params[5] || 0
  quizUUID = this.params[4]

  submitQuizModal = () => this.props.loadModal(SUBMIT_QUIZ_MODAL)

  reportQuestionModal = () => this.props.loadModal(REPORT_QUESTION_MODAL)

  // will take you to the previous question and update answer array
  // if a choice is selected
  public previousQuizQuestion = (e: any) => {
    e.preventDefault()
    this.props.changeQuestionNumber(this.props.questionNumber - 1)
  }

  // will take you to the next question and update answer array
  // if a choice is selected
  public nextQuizQuestion = (e: any) => {
    e.preventDefault()
    this.props.changeQuestionNumber(this.props.questionNumber + 1)
  }

  // choice === current 1 in 4 option
  handleSelection = (choice, title) => {
    const { quizResults } = this.props

    const quizTitles = quizResults.map(question => question.title)

    const quizQuestion = quizResults.find(question => question.title === title)

    console.log('on question ', quizQuestion)
    console.log('choice ', choice)

    const rightAnswer = quizQuestion.answers.some(
      answer =>
        answer.answer === choice &&
        parseInt(answer.percentPoints) > 0 &&
        answer.selected
    )

    if (rightAnswer) return 'choice selected-and-correct'

    const wrongAnswer = quizQuestion.answers.some(
      answer =>
        answer.answer === choice &&
        parseInt(answer.percentPoints) === 0 &&
        answer.selected
    )

    if (wrongAnswer) return 'choice selected-and-incorrect'

    const theAnswer = quizQuestion.answers.some(
      answer =>
        answer.answer === choice &&
        parseInt(answer.percentPoints) > 0 &&
        !answer.selected
    )
    if (theAnswer) return 'choice correct'

    if (!theAnswer && !rightAnswer && !wrongAnswer) return 'choice'
  }

  // const final = quizQuestion.answers.map(answer => {
  //   if (
  //     answer.answer === choice &&
  //     parseInt(answer.percentPoints) > 0 &&
  //     answer.selected
  //   )
  //     return 'choice selected-and-correct'

  //   if (
  //     answer.answer === choice &&
  //     parseInt(answer.percentPoints) === 0 &&
  //     answer.selected
  //   )
  //     return 'choice selected-and-incorrect'

  //   if (
  //     answer.answer === choice &&
  //     parseInt(answer.percentPoints) &&
  //     !answer.selected
  //   )
  //     return 'choice'
  // })

  // console.log('final ', final)
  // return final
  // const classes = quizResults.map(
  //   (question, i) => {
  //     // question.title === title &&

  //     console.log('checking answer of: ', i, question.answer)
  //     // if (!answer.selected) return 'choice'

  //     if (parseInt(answer.percentPoints) > 0 && !answer.selected)
  //       return 'choice correct'

  //     if (parseInt(answer.percentPoints) > 0 && answer.selected)
  //     return 'choice selected-and-correct'

  //     if (parseInt(answer.percentPoints) === 0 && answer.selected)
  //       return (classNames = 'choice selected-and-incorrect')
  //   }
  // )
  // console.log('classes array', classes)
  // return classes
  // if (classNames.includes('choice selected-and-correct'))
  //   return 'choice selected-and-correct'

  // if (classNames.includes('choice correct')) return 'choice correct'

  // if (classNames.includes('choice selected-and-incorrect'))
  //   return 'choice selected-and-incorrect'

  // if (
  //   !classNames.includes('choice selected-and-incorrect') &&
  //   !classNames.includes('choice selected-and-correct') &&
  //   !classNames.includes('choice correct')
  // )
  //   return 'choice'
  // }
  // handleSelection = (choice, title) => {
  //   const { quizResults } = this.props

  //   // state.takeQuiz.results.questions
  //   const quizTitles = quizResults.map(question => question.title)

  //   /* check previous answers */
  //   const aarr = quizTitles.some(ans => quizTitles.includes(title))

  //   let classNames = 'choice'
  //   const classes = quizResults.map(
  //     question =>
  //       question.title === title &&
  //       question.answers.map(answer => {
  //         // if (question.format === 'true-false') {
  //         // console.log('now in true false')
  //         // console.log('titles', question.title, title)
  //         console.log('current answer', question.answer)
  //         if (!answer.selected) return 'choice'

  //         if (parseInt(answer.percentPoints) > 0 && !answer.selected)
  //           return 'choice correct'

  //         if (parseInt(answer.percentPoints) > 0 && answer.selected)
  //           return (classNames = 'choice selected-and-correct')

  //         if (parseInt(answer.percentPoints) === 0 && answer.selected)
  //           return (classNames = 'choice selected-and-incorrect')
  //       })
  //   )
  //   console.log('classes array', classes)
  //   // return classes
  //   if (classNames.includes('choice selected-and-correct'))
  //     return 'choice selected-and-correct'

  //   if (classNames.includes('choice correct')) return 'choice correct'

  //   if (classNames.includes('choice selected-and-incorrect'))
  //     return 'choice selected-and-incorrect'

  //   if (
  //     !classNames.includes('choice selected-and-incorrect') &&
  //     !classNames.includes('choice selected-and-correct') &&
  //     !classNames.includes('choice correct')
  //   )
  //     return 'choice'
  // }

  //   if (question.format === 'multiple-choice') {
  //     console.log('in multiple choice')

  //     return (classNames =
  //       // const multArr =
  //       question.title === title &&
  //       question.answers.map(answer => {
  //         if (!answer.selected) return (classNames = 'choice')

  //         if (parseInt(answer.percentPoints) > 0 && answer.selected)
  //           return (classNames = 'choice selected-and-correct')

  //         if (parseInt(answer.percentPoints) === 0 && answer.selected)
  //           return (classNames = 'choice selected-and-incorrect')

  //         if (parseInt(answer.percentPoints) > 0 && !answer.selected)
  //           return (classNames = 'choice correct')

  //         console.log('class name is now:', classNames)
  //       }))
  //     // console.log('multArr', multArr)
  //   }

  //   if (question.format === 'multiple-select') {
  //     console.log('now in multi select')
  //     return (classNames =
  //       question.title === title &&
  //       question.answers.map(answer => {
  //         if (!answer.selected) return (classNames = ' choice')

  //         if (parseInt(answer.percentPoints) > 0 && answer.selected)
  //           return (classNames = 'choice selected-and-correct')

  //         if (parseInt(answer.percentPoints) === 0 && answer.selected)
  //           return (classNames = 'choice selected-and-incorrect')

  //         console.log('class names is now', classNames)
  //       }))
  //   }
  // })
  // console.log('classNames array', classNames)
  // // return classNames

  // if (classNames.includes('choice selected-and-correct'))
  //   return 'choice selected-and-correct'

  // if (classNames.includes('choice correct')) return 'choice correct'

  // if (classNames.includes('choice selected-and-incorrect'))
  //   return 'choice selected-and-incorrect'

  // if (
  //   !classNames.includes('choice selected-and-incorrect') &&
  //   !classNames.includes('choice selected-and-correct') &&
  //   !classNames.includes('choice correct')
  // )
  // return 'choice'
  // })
  // handleSelection = (choice, title) => {
  //   const { quizResults } = this.props

  //   // state.takeQuiz.results.questions
  //   const quizTitles = quizResults.map(question => question.title)

  //   /* check previous answers */
  //   const aarr = quizTitles.some(ans => quizTitles.includes(title))

  //   let classNames = 'choice'
  //   const classes = quizResults.map(question => {
  //     if (question.format === 'true-false') {
  //       console.log('now in true false')
  //       console.log('titles', question.title, title)
  //       question.title === title &&
  //         question.answers.map(answer => {
  //           if (!answer.selected) {
  //             classNames = ' choice'
  //           }

  //           if (parseInt(answer.percentPoints) > 0 && !answer.selected) {
  //             classNames = 'choice correct'
  //           }

  //           if (answer.percentPoints > 0 && answer.selected) {
  //             classNames = 'choice selected-and-correct'
  //           }

  //           if (parseInt(answer.percentPoints) === 0 && answer.selected) {
  //             classNames = 'choice selected-and-incorrect'
  //           }

  //           console.log('classnames aftermap', classNames)
  //         })
  //     }

  //     if (question.format === 'multiple-choice') {
  //       console.log('in multiple choice')

  //       return (classNames =
  //         // const multArr =
  //         question.title === title &&
  //         question.answers.map(answer => {
  //           if (!answer.selected) return (classNames = 'choice')

  //           if (parseInt(answer.percentPoints) > 0 && answer.selected)
  //             return (classNames = 'choice selected-and-correct')

  //           if (parseInt(answer.percentPoints) === 0 && answer.selected)
  //             return (classNames = 'choice selected-and-incorrect')

  //           if (parseInt(answer.percentPoints) > 0 && !answer.selected)
  //             return (classNames = 'choice correct')

  //           console.log('class name is now:', classNames)
  //         }))
  //       // console.log('multArr', multArr)
  //     }

  //     if (question.format === 'multiple-select') {
  //       console.log('now in multi select')
  //       return (classNames =
  //         question.title === title &&
  //         question.answers.map(answer => {
  //           if (!answer.selected) return (classNames = ' choice')

  //           if (parseInt(answer.percentPoints) > 0 && answer.selected)
  //             return (classNames = 'choice selected-and-correct')

  //           if (parseInt(answer.percentPoints) === 0 && answer.selected)
  //             return (classNames = 'choice selected-and-incorrect')

  //           console.log('class names is now', classNames)
  //         }))
  //     }
  //   })
  //   console.log('classNames array', classNames)
  //   // return classNames

  //   if (classNames.includes('choice selected-and-correct'))
  //     return 'choice selected-and-correct'

  //   if (classNames.includes('choice correct')) return 'choice correct'

  //   if (classNames.includes('choice selected-and-incorrect'))
  //     return 'choice selected-and-incorrect'

  //   if (
  //     !classNames.includes('choice selected-and-incorrect') &&
  //     !classNames.includes('choice selected-and-correct') &&
  //     !classNames.includes('choice correct')
  //   )
  //     return 'choice'
  // }

  // handleSelection = (choice, title) => {
  //   const {
  //     multipleChoiceAnswer: multCh,
  //     multipleSelectAnswer: multSe,
  //     answerArray: ansArr,
  //     questionNumber: curr,
  //     quiz
  //   } = this.props

  //   const quizTitles = quiz.questions.map(question => question.title)

  //   /*birth check*/
  //   const mca = multCh !== null && multCh !== undefined

  //   const msa =
  //     multCh === null && multSe !== undefined && multSe.answer.length > 0

  //   /* check previous answers */
  //   const aarr =
  //     !mca &&
  //     multSe !== undefined &&
  //     multSe.answer.length === 0 &&
  //     ansArr.length > 0 &&
  //     ansArr.some(ans => quizTitles.includes(title))
  //   // ansArr.length >= curr
  //   // ansArr[curr] !== undefined

  //   /* display defaults */
  //   const choose =
  //     !mca &&
  //     multSe !== undefined &&
  //     multSe.answer.length === 0 &&
  //     ansArr.every(ans => !quizTitles.includes(title))

  //   // check multiple choice object
  //   if (mca) {
  //     console.log('checking multiple choice...')
  //     return multCh.answer === choice ? 'choice selected' : 'choice'
  //   }

  //   // check multiple select object
  //   if (msa) {
  //     // if (msa && multSe.answers !== undefined) {
  //     console.log(
  //       'checking multiple select...',
  //       multSe.answer.some(answer => answer === choice)
  //     )
  //     return multSe.answer.some(answer => answer === choice)
  //       ? multSe.answer
  //           .map(answer => {
  //             if (answer === choice) return 'choice selected'
  //           })
  //           .join('')
  //       : 'choice'
  //   }

  //   // check answer array
  //   if (aarr) {
  //     // console.log('here is answers array', ansArr)
  //     const arr = ansArr.map(answer => {
  //       // let finalClass = ansArr.map(answer => {
  //       console.log('logging tritle? ', answer.title)
  //       let className = ''
  //       if (!Array.isArray(answer)) {
  //         if (answer.answer === choice && answer.title === title) {
  //           className = 'choice selected'
  //           console.log('not an array, className: ', className)
  //         }
  //         if (answer.answer !== choice) {
  //           className = 'choice'
  //           console.log('not an array, className: ', className)
  //         }
  //       }
  //       if (Array.isArray(answer.answer) && answer.title === title) {
  //         answer.answer.includes(choice)
  //           ? (className = 'choice selected')
  //           : (className = 'choice')
  //         console.log('is an array, className is now: ', className)
  //       }
  //       console.log('before submission, className is', className)
  //       return className
  //     })

  //     return arr.includes('choice selected') ? 'choice selected' : 'choice'
  //   }

  //   if (choose) {
  //     return 'choice'
  //   }
  // }

  // @ts-ignore
  render = () => {
    const { quiz, questionNumber, quizAttempt } = this.props
    return (
      <div className="review-quiz-page">
        {quiz !== null && quiz === undefined
          ? quiz.questions === undefined && (
              <Spinner
                className="loading-indicator"
                name="ball-spin-fade-loader"
              />
            )
          : quizAttempt === undefined && (
              <Spinner
                className="loading-indicator"
                name="ball-spin-fade-loader"
              />
            )}
        {quiz !== null
          ? quiz.questions !== undefined && (
              <div className="main">
                <header>
                  <div className="meta">
                    <div className="container">
                      <p className="current">Question {questionNumber + 1}</p>
                      <div className="icon" onClick={this.reportQuestionModal}>
                        <FontAwesomeIcon icon="ellipsis-h" className="menu" />
                      </div>
                    </div>
                    <p className="title">{quiz.title}</p>
                    <p className="progress">
                      Question {questionNumber + 1}/{quiz.questions.length}
                    </p>
                  </div>
                  <p className="question">
                    {quiz.questions[questionNumber].title}
                  </p>
                </header>
                {/* DISPLAYS THE CHOICES */}
                <main>
                  <div className="choices">
                    {quiz.questions[questionNumber].answers.map(
                      (answers, i) => (
                        <div
                          key={i}
                          // key={answers.answer.answer}

                          className={this.handleSelection(
                            answers.answer,
                            quiz.questions[questionNumber].title
                          )}
                          // className="choice"
                        >
                          <p>{answers.answer}</p>
                        </div>
                      )
                    )}
                  </div>
                  <div className="image-gallery-div">
                    {quiz.questions[questionNumber].image !== undefined && (
                      <ImageGallery
                        items={[
                          {
                            original: quiz.questions[questionNumber].image,
                            thumbnail: quiz.questions[questionNumber].image
                          }
                        ]}
                        showThumbnails={false}
                        showPlayButton={false}
                        showImageFullScreenButton
                        alt="question image"
                      />
                    )}
                  </div>
                  <div className="buttons">
                    {questionNumber !== 0 && (
                      <button
                        onClick={this.previousQuizQuestion}
                        className="previous-button"
                      >
                        previous
                      </button>
                    )}
                    {questionNumber + 1 !== quiz.questions.length ? (
                      <button
                        onClick={this.nextQuizQuestion}
                        className="next-button"
                      >
                        next
                      </button>
                    ) : (
                      <Link to={`/view-quiz/${this.quizUUID}`} className="link">
                        <p className="retake-button">Retake quiz</p>
                      </Link>
                    )}
                  </div>
                </main>
              </div>
            )
          : quizAttempt !== undefined &&
            quizAttempt[this.retakeIndex].questions !== undefined && (
              <div className="main">
                <header>
                  <div className="meta">
                    <div className="container">
                      <p className="current">Question {questionNumber + 1}</p>
                      <div className="icon" onClick={this.reportQuestionModal}>
                        <FontAwesomeIcon icon="ellipsis-h" className="menu" />
                      </div>
                    </div>
                    <p className="title">
                      {quizAttempt[this.retakeIndex].title}
                    </p>
                    <p className="progress">
                      Question {questionNumber + 1}/
                      {quizAttempt[this.retakeIndex].questions.length}
                    </p>
                  </div>
                  <p className="question">
                    {
                      quizAttempt[this.retakeIndex].questions[questionNumber]
                        .title
                    }
                  </p>
                </header>
                {/* DISPLAYS THE CHOICES */}
                <main>
                  <div className="choices">
                    {/*map quiz Attempts here*/}
                    {quizAttempt[this.retakeIndex].questions[
                      questionNumber
                    ].answers.map((answers, i) => (
                      <div
                        key={i}
                        // key={answers.answer.answer}
                        className="choice"
                      >
                        <p>{answers.answer}</p>
                      </div>
                    ))}
                  </div>
                  <div className="buttons">
                    {questionNumber !== 0 && (
                      <button
                        onClick={this.previousQuizQuestion}
                        className="previous-button"
                      >
                        previous
                      </button>
                    )}
                    {questionNumber + 1 !==
                    quizAttempt[this.retakeIndex].questions.length ? (
                      <button
                        onClick={this.nextQuizQuestion}
                        className="next-button"
                      >
                        next
                      </button>
                    ) : (
                      <Link to={`/view-quiz/${this.quizUUID}`} className="link">
                        <p className="retake-button">Retake quiz</p>
                      </Link>
                    )}
                  </div>
                </main>
              </div>
            )}
      </div>
    )
  }
}

const mapStateToProps = (state, props) => ({
  username: state.auth.username,
  quiz: state.takeQuiz.results,
  quizAttempt: state.quizzes.allAttempts,
  quizResults:
    state.takeQuiz.results !== null && state.takeQuiz.results.questions,
  answers: state.takeQuiz.answers,
  answerArray: state.takeQuiz.answerArray,
  questionNumber: state.takeQuiz.questionNumber,
  multipleSelectAnswer: state.takeQuiz.multipleSelectAnswer,
  multipleChoiceAnswer: state.takeQuiz.multipleChoiceAnswer,
  done: state.takeQuiz.done
})

export default connect(
  mapStateToProps,
  {
    startAddAnswerToArray,
    changeQuestionNumber,
    addMultipleChoiceAnswer,
    addMultipleSelectAnswer,
    updateMultipleSelectAnswer,
    updateAnswerArray,
    submitQuizAttempt,
    loadModal,
    clearQuizAttempt
  }
)(QuizAttemptReviewPage)
