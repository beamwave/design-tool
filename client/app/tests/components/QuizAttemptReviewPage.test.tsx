import React, { Component } from 'react'
import { shallow } from 'enzyme'
import { QuizAttemptReviewPage } from '../../components/QuizAttemptReviewPage'

/* make mock data for all that */
/* make mock functions for all those */

test('should render QuizAttemptReviewPage component', () => {
  const wrapper = shallow(
    <QuizAttemptReviewPage
      username="limp_hashbrown"
      quiz="placeholder quiz"
      quizAttempt="placeholder attempt"
      questionNumber={4}
      multipleSelectAnswer="placeholder answer"
      quizResults={['array of quizzes']}
      multipleChoiceAnswer={['placeholder array of multiple choice']}
      answerArray={['placeholder array of answers']}
      loadModal={() => `placeholder function`}
      handleSelection={() => `placeholder function`}
      clearQuizAttempt={() => `placeholder function`}
      changeQuestionNumber={() => `placeholder function`}
      startAddAnswerToArray={() => `placeholder function`}
      addMultipleChoiceAnswer={() => `placeholder function`}
      addMultipleSelectAnswer={() => `placeholder function`}
      updateMultipleSelectAnswer={() => `placeholder function`}
      updateAnswerArray={() => `placeholder function`}
      submitQuizAttempt={() => `placeholder function`}
    />
  )
  expect(wrapper).toMatchSnapshot()
})
