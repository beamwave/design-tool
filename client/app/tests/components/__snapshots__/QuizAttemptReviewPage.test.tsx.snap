// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render QuizAttemptReviewPage component 1`] = `
<div
  className="review-quiz-page"
/>
`;
